src/cpp/utils/thread_impl/thread_impl_pthread.ipp

```diff
@@ -0,0 +1,213 @@
+        // Call the trampoline
+        (*Callee)();
+        // Finish thread
+        return nullptr;
+    }
+
+public:
+
+    using native_handle_type = pthread_t;
+    using id = pthread_t;
+
+    thread()
+        : thread_hnd_(native_handle_type())
+    {
+    }
+
+    template<class _Fn>
+    thread(
+            int32_t stack_size,
+            _Fn&& f)
+    {
+        int errnum;
+
+        // Prepare trampoline to pass to ThreadProxy
+        using CalleeType = typename std::decay<_Fn>::type;
+        std::unique_ptr<CalleeType> callee(new CalleeType(std::forward<_Fn>(f)));
+
+        // Construct the attributes object.
+        pthread_attr_t attr;
+        if ((errnum = ::pthread_attr_init(&attr)) != 0)
+        {
+            throw std::system_error(errnum, std::system_category(), "pthread_attr_init failed");
+        }
+
+        // Ensure the attributes object is destroyed
+        auto attr_deleter = [](pthread_attr_t* a)
+                {
+                    int err;
+                    if ((err = ::pthread_attr_destroy(a)) != 0)
+                    {
+                        throw std::system_error(err, std::system_category(), "pthread_attr_destroy failed");
+                    }
+                };
+        std::unique_ptr<pthread_attr_t, decltype(attr_deleter)> attr_scope_destroy(&attr, attr_deleter);
+
+        // Set the requested stack size, if given.
+        if (stack_size > 0)
+        {
+            if (sizeof(unsigned) <= sizeof(int32_t) &&
+                    stack_size > static_cast<int32_t>(std::numeric_limits<unsigned>::max() / 2))
+            {
+                throw std::invalid_argument("Cannot cast stack_size into unsigned");
+            }
+
+            if ((errnum = ::pthread_attr_setstacksize(&attr, stack_size)) != 0)
+            {
+                throw std::system_error(errnum, std::system_category(), "pthread_attr_setstacksize failed");
+            }
+        }
+
+        // Construct and execute the thread.
+        pthread_t hnd;
+        if ((errnum = ::pthread_create(&hnd, &attr, ThreadProxy<CalleeType>, callee.get())) != 0)
+        {
+            throw std::system_error(errnum, std::system_category(), "pthread_create failed");
+        }
+
+        thread_hnd_ = hnd;
+        if (thread_hnd_ != native_handle_type())
+        {
+            // Thread has been correctly created. Since the ThreadProxy will
+            // take ownership of the trampoline, we need to release ownership here
+            callee.release();
+        }
+    }
+
+    ~thread()
+    {
+        if (joinable())
+        {
+            std::terminate();
+        }
+    }
+
+    // *INDENT-OFF*
+    thread(const thread&) = delete;
+    thread& operator =(const thread&) = delete;
+    // *INDENT-ON*
+
+    thread(
+            thread&& other) noexcept
+        : thread_hnd_(native_handle_type())
+    {
+        std::swap(thread_hnd_, other.thread_hnd_);
```

Comment:
Maybe use the local `swap`

Reply:
35c20702c9b0f03cd44cb144d9b3608c130aabf9